store.go

```diff
@@ -1,199 +1,207 @@
+	ErrEmptyID = errors.New("a record without id is not allowed")
 )
 
+// Store is a structure capable of retrieving records from a concrete table in
+// the database.
 type Store struct {
-	db         *mgo.Database
-	collection string
+	db       squirrel.DBProxy
+	schema   Schema
+	inserter squirrel.InsertBuilder
+	updater  squirrel.UpdateBuilder
+	deleter  squirrel.DeleteBuilder
 }
 
-// NewStore returns a new Store instance
-func NewStore(db *mgo.Database, collection string) *Store {
+// NewStore returns a new Store instance.
+func NewStore(db *sql.DB, schema Schema) *Store {
+	builder := squirrel.StatementBuilder.PlaceholderFormat(squirrel.Dollar)
+	proxy := squirrel.NewStmtCacher(db)
 	return &Store{
-		db:         db,
-		collection: collection,
+		db:       proxy,
+		schema:   schema,
+		inserter: builder.Insert(schema.Table()).RunWith(proxy),
+		updater:  builder.Update(schema.Table()).RunWith(proxy),
+		deleter:  builder.Delete(schema.Table()).RunWith(proxy),
 	}
 }
 
-// Insert insert the given document in the collection, returns error if no-new
-// document is given. The document id is setted if is empty.
-func (s *Store) Insert(doc DocumentBase) error {
-	if !doc.IsNew() {
+// Insert insert the given record in the table, returns error if no-new
+// record is given. The record id is set if it's empty.
+func (s *Store) Insert(record Record) error {
+	if !record.IsPersisted() {
 		return ErrNonNewDocument
 	}
 
-	if len(doc.GetId()) == 0 {
-		doc.SetId(bson.NewObjectId())
+	if record.GetID().IsEmpty() {
+		record.SetID(NewID())
 	}
 
-	sess, c := s.getSessionAndCollection()
-	defer sess.Close()
-
-	err := c.Insert(doc)
-	if err == nil {
-		doc.SetIsNew(false)
+	cols := s.schema.Columns()
+	values, err := RecordValues(record, cols...)
+	if err != nil {
+		return err
 	}
 
-	return err
-}
-
-// Update update the given document in the collection, returns error if a new
-// document is given.
-func (s *Store) Update(doc DocumentBase) error {
-	if doc.IsNew() {
-		return ErrNewDocument
+	_, err = s.inserter.
+		Columns(cols...).
+		Values(values...).
+		Exec()
+	if err != nil {
+		return err
 	}
 
-	sess, c := s.getSessionAndCollection()
-	defer sess.Close()
-
-	return c.UpdateId(doc.GetId(), doc)
+	record.setPersisted(false)
+	return nil
 }
 
-// Save insert or update the given document in the collection, a document with
-// id should be provided. Upsert is used (http://godoc.org/gopkg.in/mgo.v2#Collection.Upsert)
-func (s *Store) Save(doc DocumentBase) (updated bool, err error) {
-	id := doc.GetId()
-	if len(id) == 0 {
-		return false, ErrEmptyID
+// Update updates the given fields of a record in the table. All fields are
+// updated if no fields are provided. For an update to take place, the record is
+// required to have a non-empty ID and not to be a new record.
+// Returns the number of updated rows and an error, if any.
+func (s *Store) Update(record Record, cols ...string) (int64, error) {
+	if record.IsPersisted() {
```

Comment:
isNew != isPersisted
then... it should be 
`if !record.IsPersisted() {
    return 0, ErrNewDocument
}`